src/adc.rs

```diff
@@ -0,0 +1,549 @@
+                ///
+                pub fn $adcx(
+                    rb: $ADC,
+                    adc_common : &mut $ADC_COMMON,
+                    ahb: &mut AHB,
+                    ckmode: CkMode,
+                    clocks: Clocks,
+                ) -> Self {
+                    let mut this_adc = Self {
+                        rb,
+                        clocks,
+                        ckmode,
+                        operation_mode: None,
+                    };
+                    if !(this_adc.clocks_welldefined(clocks)) {
+                        panic!("Clock settings not well defined");
+                    }
+                    if !(this_adc.enable_clock(ahb, adc_common)){
+                        panic!("Clock already enabled with a different setting");
+                    }
+                    this_adc.set_align(Align::default());
+                    this_adc.calibrate();
+                    // ADEN bit cannot be set during ADCAL=1
+                    // and 4 ADC clock cycle after the ADCAL
+                    // bit is cleared by hardware
+                    this_adc.wait_adc_clk_cycles(4);
+                    this_adc.enable();
+
+                    this_adc
+                }
+
+                /// Software can use CkMode::SYNCDIV1 only if
+                /// hclk and sysclk are the same. (see reference manual 15.3.3)
+                fn clocks_welldefined(&self, clocks: Clocks) -> bool {
+                    if (self.ckmode == CkMode::SYNCDIV1)
+                    {
+                        clocks.hclk().0 == clocks.sysclk().0
+                    } else {
+                        true
+                    }
+                }
+
+                /// sets up adc in one shot mode for a single channel
+                pub fn setup_oneshot(&mut self) {
+                    self.rb.cr.modify(|_, w| w.adstp().stop());
+                    self.rb.isr.modify(|_, w| w.ovr().clear());
+
+                    self.rb.cfgr.modify(|_, w| w
+                        .cont().single()
+                        .ovrmod().preserve()
+                    );
+
+                    self.set_sequence_len(1);
+
+                    self.operation_mode = Some(OperationMode::OneShot);
+                }
+
+                fn set_sequence_len(&mut self, len: u8) {
+                    debug_assert!(len <= 16);
+                    self.rb.sqr1.modify(|_, w| w.l().bits(len - 1));
+                }
+
+                fn set_align(&self, align: Align) {
+                    self.rb.cfgr.modify(|_, w| w.align().variant(align.into()));
+                }
+
+                fn enable(&mut self) {
+                    self.rb.cr.modify(|_, w| w.aden().enable());
+                    while self.rb.isr.read().adrdy().is_not_ready() {}
+                }
+
+                fn disable(&mut self) {
+                    self.rb.cr.modify(|_, w| w.aden().clear_bit());
```

Comment:
I haven't actually checked this, but I'm pretty sure there is no `disable` method, because you cannot disable the ADC via the `ADEN` bit. You need to use `ADDIS`, which also has a `disable` method.